File: siggen.py
#!/usr/bin/env python
# -*- coding: iso-8859-1 -*-
#
# $Id: siggen.py,v 1.2 2006/05/30 21:47:04 grahn Exp $
#
# Copyright (c) 2006 Jörgen Grahn.
# All rights reserved.
"""formatting texts into filler parts of a sig file
"""
import re
import random

def _lentab(ss):
    nn = []
    n = 0
    for s in ss:
        n += len(s)
        nn.append(n)
        n += 1
    return nn

class Subs:
    def __init__(self, s):
        self.ss = s.split()
        self.nn = _lentab(self.ss)
    def consume(self, n):
        parts = 0
        for k in self.nn:
            if k > n: break
            parts += 1
        # include the first 'parts' of self.ss
        res = ' '.join(self.ss[:parts])
        del self.ss[:parts]
        self.nn = _lentab(self.ss)
        return res + ' ' * (n - len(res))

def fill(template, fortune):
    """
    """
    pat = re.compile(r'(#+)')
    subs = Subs(fortune)
    acc = []
    for t in pat.split(template):
        if not t.startswith('#'):
            acc.append(t)
            continue
        n = len(t)
        acc.append(subs.consume(n))
    return ''.join(acc)

class Fortunes:
    def __init__(self, f):
        lines = f.readlines()
        random.shuffle(lines)
        self._lines = lines
    def get(self):
        s = self._lines[0]
        del self._lines[0]
        return s

if __name__ == "__main__":
    import sys
    import os
    import getopt

    prog = os.path.split(sys.argv[0])[1]
    usage = 'usage: %s -f quotes-file signature-file' % prog
    try:
        opts, argv = getopt.getopt(sys.argv[1:], 'f:')
        fortunes = None
        for option, value in opts:
            if option == '-f': fortunes = value
        if not fortunes:
            raise ValueError('required argument missing')
        template, = argv
    except (ValueError, getopt.GetoptError), s:
        print >>sys.stderr, 'error:', s
        print >>sys.stderr, usage
        sys.exit(1)

    fortunes = Fortunes(open(fortunes, 'r'))
    template = open(template, 'r').read()
    sys.stdout.write(fill(template, fortunes.get()))

File: siggen.cc
/*
 * Copyright (c) 2024 Jörgen Grahn
 * All rights reserved.
 *
 */
#include <iostream>
#include <string>
#include <vector>
#include <fstream>
#include <cstring>

#include <getopt.h>

#include "fortune.h"
#include "render.h"

namespace {

    /* Read a whole file (hopefully a shortish one) into a string.
     */
    std::string read(std::istream& is)
    {
	char buf[10];
	std::string s;
	while (1) {
	    is.read(buf, sizeof buf);
	    s.append(buf, is.gcount());
	    if (!is) break;
	}
	return s;
    }
}

int main(int argc, char ** argv)
{
    const std::string prog = argv[0] ? argv[0] : "siggen";
    const std::string usage = "usage: "
	+ prog +
	" -f quotes ..."
	" signature\n"
	"       "
	+ prog + " --help\n" +
	"       "
	+ prog + " --version";
    const char optstring[] = "f:";
    const struct option long_options[] = {
	{"help",    	 0, 0, 'h'},
	{"version", 	 0, 0, 'v'},
	{0, 0, 0, 0}
    };

    std::vector<std::string> qpaths;
    std::string spath;

    int ch;
    while ((ch = getopt_long(argc, argv,
			     optstring, &long_options[0], 0)) != -1) {
	switch(ch) {
	case 'f':
	    qpaths.push_back(optarg);
	    break;
	case 'h':
	    std::cout << usage << '\n';
	    return 0;
	case 'v':
	    std::cout << "siggen 2.0\n"
		      << "Copyright (c) 2006 - 2024 Jörgen Grahn\n";
	    return 0;
	    break;
	case ':':
	case '?':
	    std::cerr << usage << '\n';
	    return 1;
	    break;
	default:
	    break;
	}
    }

    if (optind + 1 == argc) spath = argv[optind];

    if (qpaths.empty() || spath.empty()) {
	std::cerr << "error: required argument missing\n"
		  << usage << '\n';
	return 1;
    }

    const std::string fs = fortune(std::cerr, qpaths);

    std::ifstream f {spath};
    if (!f) {
	std::cerr << "error: cannot open " << spath
		  << "for writing: " << std::strerror(errno) << '\n';
	return 1;
    }

    const auto src = read(f);
    std::cout << render(src, fs);

    return 0;
}
